attach/mutt_attach.c

```diff
@@ -366,6 +378,185 @@ void mutt_check_lookup_list(struct Body *b, char *type, size_t len)
   }
 }
 
+/**
+ * save_cid_attachment - Save attachment if it has a Content-ID
+ * @param[in]  body         Body to check and save
+ * @param[out] cid_map_list List of Content-ID to filename mappings
+ *
+ * If body has a Content-ID, it is saved to disk and a new Content-ID to filename
+ * mapping is added to cid_map_list.
+ */
+static void save_cid_attachment(struct Body *body, struct CidMapList *cid_map_list)
+{
+  if (!body || !cid_map_list)
+    return;
+
+  char *id = mutt_param_get(&body->parameter, "content-id");
+  struct Buffer *tmpfile = mutt_buffer_pool_get();
+  struct Buffer *cid = mutt_buffer_pool_get();
+  bool has_tempfile = false;
+  FILE *fp = NULL;
+
+  if (id)
+  {
+    mutt_debug(LL_DEBUG2, "attachment found with \"Content-ID: %s\"\n", id);
+    /* Get filename */
+    char *fname = mutt_str_dup(body->filename);
+    if (body->aptr)
+    {
+      fp = body->aptr->fp;
+    }
+    mutt_file_sanitize_filename(fname, fp ? true : false);
+    mailcap_expand_filename("%s", fname, tmpfile);
+    FREE(&fname);
+
+    /* Save attachment */
+    if (mutt_save_attachment(fp, body, mutt_buffer_string(tmpfile), 0, NULL) == -1)
+      goto bail;
+    has_tempfile = true;
+    mutt_debug(LL_DEBUG2, "attachment with \"Content-ID: %s\" saved to file \"%s\"\n",
+               id, mutt_buffer_string(tmpfile));
+
+    /* Add Content-ID to filename mapping to list */
+    struct CidMap *cid_map = mutt_mem_malloc(sizeof(struct CidMap));
+    mutt_buffer_printf(cid, "cid:%s", id);
+    cid_map->cid = mutt_str_dup(mutt_buffer_string(cid));
+    cid_map->fname = mutt_str_dup(mutt_buffer_string(tmpfile));
```

Comment:
May I suggest a `struct CidMap *cid_map_new(const char *cid, const char *tmpfile)` function?

Reply:
Seconded.
And a `_free()` function (as it's non-trivial)
And `calloc`, to be safe